File: Tests/Strata.Foundation.CoreTest/test/typescript/Utility/ConditionalTest.ts
import "jest"
import {Conditional}  from "strata.foundation.core/Utility/Conditional";
import {Holder, NoSuchElementException} from "strata.foundation.core/Utility";

describe(
    "Conditional",
    () =>
    {
        it(
            "testIfTrueSupplier",
            () =>
            {
                const trueCondition: Conditional = Conditional.TRUE;
                const falseCondition: Conditional = Conditional.FALSE;
                const expected: string = "condition is true";

                trueCondition
                    .ifTrue(() => expected)
                    .ifPresentOrElseNoReturn(
                        actual => expect(actual).toBe(expected),
                        () => {throw new NoSuchElementException("");});
                falseCondition
                    .ifTrue(() => expected)
                    .ifPresentOrElseNoReturn(
                        actual => {throw new Error(actual);},
                        () => {});
            });
        it(
            "testIfFalseSupplier",
            () =>
            {
                const trueCondition: Conditional = Conditional.TRUE;
                const falseCondition: Conditional = Conditional.FALSE;
                const expected: string = "condition is true";

                falseCondition
                    .ifFalse(() => expected)
                    .ifPresentOrElseNoReturn(
                        actual => expect(actual).toBe(expected),
                        () => {throw new NoSuchElementException("");});
                trueCondition
                    .ifFalse(() => expected)
                    .ifPresentOrElseNoReturn(
                        actual => {throw new Error(actual);},
                        () => {});
            });
        it(
            "testIfTrueRunnable",
            () =>
            {
                const trueCondition: Conditional = Conditional.TRUE;
                const falseCondition: Conditional = Conditional.FALSE;
                const expected: string = "condition is true";
                const actual: Holder<string> = new Holder<string>();

                trueCondition
                    .ifTrueNoReturn(() => actual.accept(expected));

                expect(actual.get()).toBe(expected);

                falseCondition
                    .ifTrueNoReturn(() => {throw new Error("should not execute")});
            });
        it(
            "testIfFalseRunnable",
            () =>
            {
                const trueCondition: Conditional = Conditional.TRUE;
                const falseCondition: Conditional = Conditional.FALSE;
                const expected: string = "condition is true";
                const actual: Holder<string> = new Holder<string>();

                falseCondition
                    .ifFalseNoReturn(() => actual.accept(expected));

                expect(actual.get()).toBe(expected);

                trueCondition
                    .ifFalseNoReturn(() => {throw new Error("should not execute")});
            });
        it(
            "testIfTrueOrElseSupplier",
            () =>
            {
                const trueCondition: Conditional = Conditional.TRUE;
                const falseCondition: Conditional = Conditional.FALSE;
                const expected: string = "condition is true";
                const error: string = "error";

                expect(
                    trueCondition
                        .ifTrueOrElse(() => expected,() => error))
                    .toBe(expected);

                expect(
                    falseCondition
                        .ifTrueOrElse(() => error,() => expected))
                    .toBe(expected);

            });
        it(
            "testIfTrueOrElseRunnable",
            () =>
            {
                const trueCondition: Conditional = Conditional.TRUE;
                const falseCondition: Conditional = Conditional.FALSE;

                trueCondition.ifTrueOrElseNoReturn(
                    () => {},
                    () => {throw new Error();});

                falseCondition.ifTrueOrElseNoReturn(
                    () => {throw new Error();},
                    () => {});
            });
        it(
            "testIfTrueOrThrowSupplier",
            () =>
            {
                const trueCondition: Conditional = Conditional.TRUE;
                const falseCondition: Conditional = Conditional.FALSE;
                const expected: string = "condition is true";

                expect(
                    trueCondition
                        .ifTrueOrThrow(() => expected,new Error()))
                    .toBe(expected);

                try
                {
                    falseCondition
                        .ifTrueOrThrow(() => expect, new Error());

                    fail("should have thrown exception");
                }
                catch (e) {}
            });
        it(
            "testIfTrueOrThrowRunnable",
            () =>
            {
                const trueCondition: Conditional = Conditional.TRUE;
                const falseCondition: Conditional = Conditional.FALSE;

                trueCondition.ifTrueOrThrowNoReturn(
                    () => {},
                    new Error());

                try
                {
                    falseCondition.ifTrueOrThrowNoReturn(
                        () => {
                        },
                        new Error());

                    fail("should have thrown exception");
                }
                catch (e) {}
            });
        it(
            "testAnd",
            () =>
            {
                const andTrue: Conditional   = Conditional.TRUE.and(Conditional.TRUE);
                const andFalse1: Conditional = Conditional.FALSE.and(Conditional.FALSE);
                const andFalse2: Conditional = Conditional.TRUE.and(Conditional.FALSE);
                const andFalse3: Conditional = Conditional.FALSE.and(Conditional.TRUE);

                expect(andTrue.get()).toBeTruthy();
                expect(andFalse1.get()).toBeFalsy();
                expect(andFalse2.get()).toBeFalsy();
                expect(andFalse3.get()).toBeFalsy();
            });
        it(
            "testOr",
            () =>
            {
                const orTrue1: Conditional = Conditional.TRUE.or(Conditional.TRUE);
                const orTrue2: Conditional = Conditional.FALSE.or(Conditional.TRUE);
                const orTrue3: Conditional = Conditional.TRUE.or(Conditional.FALSE);
                const orFalse: Conditional = Conditional.FALSE.or(Conditional.FALSE);

                expect(orTrue1.get()).toBeTruthy();
                expect(orTrue2.get()).toBeTruthy();
                expect(orTrue3.get()).toBeTruthy();
                expect(orFalse.get()).toBeFalsy();
            });
        it(
            "testXor",
            () =>
            {
                const xorTrue1: Conditional  = Conditional.TRUE.xor(Conditional.FALSE);
                const xorTrue2: Conditional  = Conditional.FALSE.xor(Conditional.TRUE);
                const xorFalse1: Conditional = Conditional.TRUE.xor(Conditional.TRUE);
                const xorFalse2: Conditional = Conditional.FALSE.xor(Conditional.FALSE);

                expect(xorTrue1.get()).toBeTruthy();
                expect(xorTrue2.get()).toBeTruthy();
                expect(xorFalse1.get()).toBeFalsy();
                expect(xorFalse2.get()).toBeFalsy();
            });
        it(
            "testNot",
            () =>
            {
                const isTrue: Conditional  = Conditional.FALSE.not();
                const isFalse: Conditional  = Conditional.TRUE.not();

                expect(isTrue.get()).toBeTruthy();
                expect(isFalse.get()).toBeFalsy();
            });

    });

File: Tests/Strata.Foundation.CoreTest/test/java/strata/foundation/core/utility/ConditionalTest.java
//////////////////////////////////////////////////////////////////////////////
// ConditionalTest.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.NoSuchElementException;

import static org.junit.jupiter.api.Assertions.*;

@Tag("CommitStage")
public
class ConditionalTest
{
    private Conditional trueCondition;
    private Conditional falseCondition;

    @BeforeEach
    public void
    setUp()
    {
        trueCondition = Conditional.of(true);
        falseCondition = Conditional.of(false);
    }

    @Test
    public void
    testIfTrueSupplier()
    {
        String expected = "condition is true";

        trueCondition
            .ifTrue(() -> expected)
            .ifPresentOrElse(
                actual -> assertEquals(expected,actual),
                () -> {throw new NoSuchElementException();});
        falseCondition
            .ifTrue(() -> expected)
            .ifPresentOrElse(
                actual -> {throw new IllegalStateException();},
                () -> {});
    }

    @Test
    public void
    testIfFalseSupplier()
    {
        String expected = "condition is true";

        falseCondition
            .ifFalse(() -> expected)
            .ifPresentOrElse(
                actual -> assertEquals(expected,actual),
                () -> {throw new NoSuchElementException();});
        trueCondition
            .ifFalse(() -> expected)
            .ifPresentOrElse(
                actual -> {throw new IllegalStateException();},
                () -> {});
    }

    @Test
    public void
    testIfTrueRunnable()
    {
        String         expected = "condition is true";
        Holder<String> actual = new Holder<>();

        trueCondition
            .ifTrue(() -> {actual.setItem(expected);});

        assertEquals(expected,actual.getItem());

        falseCondition
            .ifTrue(() -> {throw new IllegalStateException();});
    }

    @Test
    public void
    testIfFalseRunnable()
    {
        String         expected = "condition is true";
        Holder<String> actual = new Holder<>();

        falseCondition
            .ifFalse(() -> {actual.setItem(expected);});

        assertEquals(expected,actual.getItem());

        trueCondition
            .ifFalse(() -> {throw new IllegalStateException();});
    }

    @Test
    public void
    testIfTrueOrElseSupplier()
    {
        String expected = "condition is true";
        String error    = "error";

        assertEquals(
            expected,
            trueCondition.ifTrueOrElse(() -> expected,() -> error));

        assertEquals(
            expected,
            falseCondition.ifTrueOrElse(() -> error,() -> expected));

    }

    @Test
    public void
    testIfTrueOrElseRunnable()
    {
        trueCondition.ifTrueOrElse(
            () -> {},
            () -> {throw new IllegalStateException();});

        falseCondition.ifTrueOrElse(
            () -> {throw new IllegalStateException();},
            () -> {});

    }

    @Test
    public void
    testIfTrueOrThrowSupplier() throws Throwable
    {
        String expected = "condition is true";

        assertEquals(
            expected,
            trueCondition.ifTrueOrThrow(
                () -> expected,
                new IllegalStateException()));

        try
        {
            falseCondition.ifTrueOrThrow(
                () -> expected,
                new IllegalStateException());

            fail("Should have thrown an exception");
        }
        catch (IllegalStateException e) {}
    }

    @Test
    public void
    testIfTrueOrThrowRunnable() throws Throwable
    {
        trueCondition.ifTrueOrThrow(
            () -> {},
            new IllegalStateException());

        try
        {
            falseCondition.ifTrueOrThrow(
                () -> {},
                new IllegalStateException());

            fail("Should have thrown an exception");
        }
        catch (IllegalStateException e) {}
    }

    @Test
    public void
    testAnd()
    {
        Conditional andTrue   = trueCondition.and(trueCondition);
        Conditional andFalse1 = falseCondition.and(falseCondition);
        Conditional andFalse2 = trueCondition.and(falseCondition);
        Conditional andFalse3 = falseCondition.and(trueCondition);

        assertTrue(andTrue.get());
        assertFalse(andFalse1.get());
        assertFalse(andFalse2.get());
        assertFalse(andFalse3.get());
    }

    @Test
    public void
    testOr()
    {
        Conditional orTrue1  = trueCondition.or(trueCondition);
        Conditional orTrue2  = trueCondition.or(falseCondition);
        Conditional orTrue3  = falseCondition.or(trueCondition);
        Conditional orFalse  = falseCondition.or(falseCondition);

        assertTrue(orTrue1.get());
        assertTrue(orTrue2.get());
        assertTrue(orTrue3.get());
        assertFalse(orFalse.get());
    }

    @Test
    public void
    testXor()
    {
        Conditional xorTrue1  = trueCondition.xor(falseCondition);
        Conditional xorTrue2  = falseCondition.xor(trueCondition);
        Conditional xorFalse1 = trueCondition.xor(trueCondition);
        Conditional xorFalse2 = falseCondition.xor(falseCondition);

        assertTrue(xorTrue1.get());
        assertTrue(xorTrue2.get());
        assertFalse(xorFalse1.get());
        assertFalse(xorFalse2.get());
    }

    @Test
    public void
    testNot()
    {
        Conditional isTrue  = falseCondition.not();
        Conditional isFalse = trueCondition.not();

        assertTrue(isTrue.get());
        assertFalse(isFalse.get());
    }

    @Test
    public void
    testOf()
    {
        Conditional isTrue = Conditional.of(fooTrue());
        Conditional isFalse = Conditional.of(fooFalse());

        assertTrue(isTrue.get());
        assertFalse(isFalse.get());

        assertEquals(Conditional.TRUE, isTrue);
        assertEquals(Conditional.FALSE, isFalse);
    }

    @Test
    public void
    testUsage()
    {
        assertEquals(
            "THIS IS TRUE",
            Conditional
                .of(fooTrue())
                .ifTrueOrElse(
                    () -> "this is true",
                    () -> "this is false")
                .toUpperCase());
        assertEquals(
            "THIS IS FALSE",
            Conditional
                .of(fooFalse())
                .ifTrueOrElse(
                    () -> "this is true",
                    () -> "this is false")
                .toUpperCase());
    }

    private boolean
    fooTrue() { return true; }

    private Boolean
    fooFalse() { return false; }
}

//////////////////////////////////////////////////////////////////////////////
